src/components/cleaner.ts

```diff
@@ -44,23 +45,42 @@ export class Cleaner {
         if (configuration.get('latex.clean.subfolder.enabled') as boolean) {
             globs = globs.map(globType => './**/' + globType)
         }
+        globs = Array.from(new Set(globs))  // remove duplicate entries
         this.extension.logger.addLogMessage(`Clean glob matched files: ${JSON.stringify({globs, outdir})}`)
-        const files = globs.map(g => glob.sync(g, {cwd: outdir}))
-            // Reduce the array of arrays to a single array containing all the files that should be deleted
-            .reduce((all, curr) => all.concat(curr), [])
-            // Resolve the absolute filepath for every file
-            .map(file => path.resolve(outdir, file))
-        for (const file of files) {
+        // All glob patterns for folders, the glob should be end with a slash (`path.sep`) and the last component should not contain `**`
+        const folderGlobs: string[] = globs.filter(globType => globType.endsWith(path.sep) && !path.basename(globType).includes('**'))
+        const matchFolderGlobs = (folder: string): boolean => folderGlobs.some(globType => minimatch(folder, globType))
+        const pathSet: Set<string> = new Set(  // unique paths
+            globs.map(g => glob.sync(g, {cwd: outdir}))
+                 // Reduce the array of arrays to a single array containing all the files that should be deleted
+                 .reduce((all, curr) => all.concat(curr), [])
+        )
+        const paths: string[] = Array.from(pathSet)
+                                     // Sort by descending dictionary order, make sure the folder is sorted after the files in it
+                                     .sort((a, b) => b.localeCompare(a))
+                                     // Pattern `abc/**` matches folder "abc" while pattern `abc/` matches "abc/"
+                                     // Remove the duplicates and keep the folder paths with tailing slashes
+                                     .filter((p) => !pathSet.has(p + path.sep))
+        // Resolve absolute path before removing anything (for symlinked paths)
```

Comment:
This cannot handle cases with clean type e.g. ['*.pdf', 'test/'] where you have a pdf in folder test/. The sort will put the folder before *.pdf, leaving a non-empty folder.

Reply:
`paths` is not the glob patterns, this is the result of a list of file paths.

> This cannot handle cases with clean type e.g. ['*.pdf', 'test/'] where you have a pdf in folder test/. The sort will put the folder before *.pdf, leaving a non-empty folder.

After sorting, the list in the above case would be:

```javascript
[
    'test/abc.pdf',
    'test/'
]
```

Reply:
Let me re-iterate my example above. Suppose we have a `main.pdf` and a `temp/section.pdf`. Then we set the file type to remove to `['*.pdf', 'temp/']`. In this case, the folder `temp/` is emptied but not removed. To do so, you need to set `['*.pdf', 'temp/', 'temp/*.pdf']`, which is against the usage of a single `*.pdf` which are always prepended with ** if the subfolder config is set to true.

Reply:
> In this case, the folder `temp/` is emptied but not removed.

I'm afraid that `temp/` is not empty when `latex.clean.subfolder.enabled` is `false`. When it is `true`, the folder will be emptied and will be removed.

Suppose the following case in https://github.com/James-Yu/LaTeX-Workshop/pull/3353#discussion_r909509868:

```console
$ tree
.
├── main.pdf
└── temp
    └── section.pdf

1 directory, 2 files
```

```jsonc
{
    "latex-workshop.latex.clean.method": "glob",
    "latex-workshop.latex.clean.fileTypes": ["temp/", "*.pdf"]  // the order of globs does not matter
}
```

- If `latex.clean.subfolder.enabled` is `false`:
    1. Glob for patterns `["temp/", "*.pdf"]`, gets:
        ```json
        ["temp/", "main.pdf"]
        ```
	2. To absolute paths and sort by descending dictionary order, gets:
        ```json
        ["<outDir>/main.pdf", "<outDir>/temp/"]
        ```
    3. Remove `main.pdf`. The cleaner will keep the folder `temp/` since it's not empty (`temp/section.pdf` is not removed).

- If `latex.clean.subfolder.enabled` is `true` (prepend `./**/` to glob patterns):
    1. Glob for patterns `["./**/temp/", "./**/*.pdf"]`, gets:
        ```json
        ["./temp/", "./main.pdf", "./temp/section.pdf"]
        ```
    2. To absolute paths and sort by descending dictionary order (note that `"<outDir>/temp/section.pdf" > "<outDir>/temp/"`), gets:
        ```json
        ["<outDir>/main.pdf", "<outDir>/temp/section.pdf", "<outDir>/temp/"]
        ```
    3. Remove all elements sequentially. `<outDir>/temp/` becomes empty after removing `<outDir>/temp/section.pdf`, then it will be removed.